File: Python/Container_With_Most_Water.py
class Solution:
    # @return an integer
    """
    Reason for moving the pointer with smaller height
    current area: smaller*|index_larger - index_smaller|
    if moving the larger ptr, new area: min(smaller, newPtr)*(|index_larger - index_smaller| - 1) <= current
    if moving the smaller ptr, new area: min(newPtr, larger)*(|index_larger - index_smaller| - 1), UNKNOWN and needs to be compared
    """
    def maxArea(self, height):
        ret = 0
        n = len(height)
        left = 0
        right = -1
        while left - right < n:
            ret = max(ret, min(height[left], height[right])*(n+right-left))
            if height[left] < height[right]: left = left + 1
            else: right = right - 1
        return ret

File: Go/Container_With_Most_Water.go
func maxArea(height []int) int {
    start,end := 0,len(height)-1
    max_area,tmp_area := 0,0
    for start < end {
        if height[start] <= height[end] {
            tmp_area = height[start]*(end-start)
            start ++
        } else {
            tmp_area = height[end]*(end-start)
            end --
        }
        if tmp_area > max_area {
            max_area = tmp_area
        }
    }
    return max_area
}
